syft/execution/type_wrapper.py

```diff
@@ -0,0 +1,261 @@
+        """
+            Function to raise an error if the build/reference function has a different number of arguments.
+
+            Params:
+                obj_type: the type of the object returned by calling .__name__ on it.
+                obj_name: the name/id of the object.
+                build: the build/reference input length.
+                call: the called input length.
+
+            Returns:
+                 None
+        """
+        raise TypeError(f"{obj_type} {obj_name} requires {build} arguments, received {call}.")
+
+    @staticmethod
+    def raise_key_missing_err(obj_type: str, obj_name: str, key: any, path: str) -> None:
+        """
+            Function to raise an error if the build/reference function has a different number of arguments.
+
+            Params:
+                obj_type: the type of the object returned by calling .__name__ on type(obj).
+                obj_name: the name/id of the object.
+                key: the key that is missing from the called dict.
+                path: the nested path to reach that obj.
+
+            Returns:
+                 None
+        """
+        raise KeyError(f"{obj_type} {obj_name} {path} does not provide the key {key}, while being build with that key.")
+
+    def input_check(self, obj_type: str, obj_name: str, args: list) -> None:
+        """
+            Method for input validation by comparing the serialized build input with the
+            current call input, following the following steps:
+                1. Input length validation - checking that build and call inputs match on length.
+                2. Verify the following nested structures: list, tuple, dict recursively. Lengths must match when
+                comparing two nested lists, tuples or dicts. If they differ, an error will be raised.
+                3. If we hit an object for which we don't support nesting, we compare types between call input and
+                build input. If they differ, a warning will be raised.
+                4. Dicts on the same nesting level on build and call input must have the same keys. If they differ, an
+                error will be raised.
+
+            Params:
+                obj_type: the type of the object returned by calling .__name__ on type(obj).
+                obj_name: the name/id of the object
+                args: the arguments to be compared with the reference/build one.
+
+            Returns:
+                None
+        """
+        def check_type_nested_structure(obj_type: str, obj_name: str, build: any, call: any, path: str) -> None:
+            iterable_supported_list = (list, tuple, dict)
+
+            if type(call) not in iterable_supported_list:
+                if not isinstance(call, build):
+                    NestedTypeWrapper.raise_typecheck_warn(obj_type, obj_name, build.__name__, type(call).__name__, path)
+                return
+
+            if type(build) != type(call):
```

Comment:
Is build a type or a value? You use `isinstance(call, build)` and type(build)` which is weird
Also, I think we can find a better naming for build and call

Reply:
It depends, if they are nested objects (dict, list, tuple), they will be objects (instances of those nested types), if not, build will be a type, call will be an obj (an instance of that type, if the input is correct).